### ISSUE TITLE
`//eslint-disable-line`'s behavior is changed in 4.7.0

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#9318]
**Tell us about your environment**

* **ESLint Version:** 4.7.0
* **Node Version:** 8.4.0
* **npm Version:** 5.4.2

**What parser (default, Babel-ESLint, etc.) are you using?**

- default

**Please show your full configuration:**

- nothing

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint-disable no-unused-vars */

var foo; // eslint-disable-line no-unused-vars
var bar;

/* eslint-enable no-unused-vars */
```

```
$ eslint test.js --rule no-unused-vars:error
```

**What did you expect to happen?**

No errors as same as `4.6.0`.

**What actually happened? Please include the actual, raw output from ESLint.**

```
~\dev\sandbox> eslint test.js --rule no-unused-vars:error

C:\Users\t-nagashima.MSS\dev\sandbox\test.js
  6:5  error  'bar' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Comment by emmettnicholas** [comment]
In case it helps, here's a different repro for seemingly the same underlying issue:

**What did you do? Please include the actual source code causing the issue.**


```
let a; // eslint-disable-line no-unused-vars
let b; // eslint-disable-line no-unused-vars
let c; // eslint-disable-line no-unused-vars
let d; // eslint-disable-line no-unused-vars
let f; // eslint-disable-line no-unused-vars
let g; // eslint-disable-line no-unused-vars
```

**What did you expect to happen?**

no errors

**What actually happened? Please include the actual, raw output from ESLint.**

`2:5  warning  'b' is defined but never used  no-unused-vars`

**Comment by not-an-aardvark** [comment]
> ```js
> /* eslint-disable no-unused-vars */
> 
> var foo; // eslint-disable-line no-unused-vars
> var bar;
> 
> /* eslint-enable no-unused-vars */
> ```

I'm not sure about this -- the `// eslint-disable-next-line` disables the rule for that one line and enables it afterwards, so it makes sense to me that the rule would be enabled after it. The problem could be solved by removing the `// eslint-disable-next-line` comment, since it's redundant otherwise.

> ```js
> let a; // eslint-disable-line no-unused-vars
> let b; // eslint-disable-line no-unused-vars
> let c; // eslint-disable-line no-unused-vars
> let d; // eslint-disable-line no-unused-vars
> let f; // eslint-disable-line no-unused-vars
> let g; // eslint-disable-line no-unused-vars
> ```

This looks like a bug to me.

**Comment by sqal** [comment]
@not-an-aardvark what about [this example](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50LWRpc2FibGUgbm8tdXNlLWJlZm9yZS1kZWZpbmUgKi9cclxuXHJcbmNsYXNzIEZvbyB7XHJcblx0YmFyKGRhdGEpIHtcclxuXHRcdHRoaXMueCA9IGRhdGFbMF07IC8vIGVzbGludC1kaXNhYmxlLWxpbmVcclxuXHRcdGJhcihkYXRhWzFdKTtcclxuXHR9XHJcbn1cclxuXHJcbm5ldyBGb28oKTtcclxuXHJcbmZ1bmN0aW9uIGJhcigpIHt9Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6MiwicHJlZmVyLWRlc3RydWN0dXJpbmciOjIsIm5vLXVzZS1iZWZvcmUtZGVmaW5lIjoyfSwiZW52Ijp7fX19) It worked fine in previous version but now i cannot disable no-use-before-define rule. However if you remove // eslint-disable-line from line 5 it works again.

**Comment by not-an-aardvark** [comment]
Hmm, that's a good point. I suppose we should treat `eslint-disable-line` comments differnetly then. They're currently processed like this:

```js
// eslint-disable-next-line
foo;

bar;


// is processed the same as:


/* eslint-disable */foo;
/* eslint-enable */
bar;
```

But it seems like that isn't accurate, because applying a global `/* eslint-enable */` after the comment seems to be unintuitive since it also gets rid of any previous `/* eslint-disable */` comments.

**Comment by mysticatea** [comment]
#9216 seemed to include a breaking change. `eslint-disable/enable` pair can be nested, but it removed the feature. I think that we should revert the change, until the next major release at least.

**Comment by gyandeeps** [comment]
> #9216 seemed to include a breaking change. eslint-disable/enable pair can be nested, but it removed the feature. I think that we should revert the change, until the next major release at least.

Wow, did not know about this feature... lol
can u share an example..

**Comment by not-an-aardvark** [comment]
https://github.com/eslint/eslint/pull/9216#discussion_r136707906

**Comment by mysticatea** [comment]
@gyandeeps see the first post of this issue :D 

As @not-an-aardvark mentioned, `eslint-disable-line` is converted to `eslint-disable/enable pair`, so the example is nested.

**Comment by not-an-aardvark** [comment]
Are people actually using the nesting aside from `eslint-disable-line` comments? I had thought that the nesting cases in https://github.com/eslint/eslint/pull/9216#discussion_r136707906 were bugs, so an alternative fix would be to update `eslint-disable-line` processing to avoid using nesting.

**Comment by gyandeeps** [comment]
i agree with @not-an-aardvark .. those test (ref https://github.com/eslint/eslint/pull/9216#discussion_r136707906) are actually incorrect. Examples explained in that comment makes sense to me.

**Comment by mysticatea** [comment]
I'm not using actually because ` eslint-plugin-eslint-comments` catches it. I have noticed this because I needed to change the plugin's logic to follow the change.

**Comment by gyandeeps** [comment]
just to be clear, the code mentioned in this issue above and also the 2nd comment code example are bugs.... but https://github.com/eslint/eslint/pull/9216#discussion_r136707906 this is slightly different.

**Comment by mysticatea** [comment]
In my perspective, those look same. One more example:

```js
/* eslint-disable no-unused-vars */

function aaa() {
    /* eslint-disable */
    var foo;
    /* eslint-enable */
}

var bar;

/* eslint-enable no-unused-vars */
```

**Comment by not-an-aardvark** [comment]
I think it makes the most sense for nesting to not be supported, and for `eslint-disable` and `eslint-disable-line` comments to be considered separately without relying on nesting. That way, the examples in https://github.com/eslint/eslint/issues/9318#issue-258221265 and https://github.com/eslint/eslint/issues/9318#issuecomment-329979578 will be considered correct), and the examples in https://github.com/eslint/eslint/pull/9216#discussion_r136707906 and https://github.com/eslint/eslint/issues/9215 will be considered incorrect.

Alternatively, we could update the implementation to support nesting again. (I don't think reverting would work well since we've made some additional changes to the code since the original refactor landed). However, this could end up being confusing, because the behavior in https://github.com/eslint/eslint/issues/9215 definitely seems incorrect to me.

**Comment by mysticatea** [comment]
I agree that #9215 is a bug. 

Hmm, I suspect that it's a breaking change because the behavior exists in a long time. On the other hand, the behavior has never documented, so semver doesn't consider that it's a breaking change.

OK, I don't oppose the way no nesting.

**Comment by not-an-aardvark** [comment]
Just as a heads-up: I probably won't have time to fix this until Sunday night, so if someone else is able to fix it earlier, that might allow for more time to review before we do a patch release on Monday.

**Comment by shellscape** [comment]
As @mysticatea [commented](https://github.com/eslint/eslint/issues/9318#issuecomment-329997336), our block-disabling comments stopped working in 4.7.0 as well. 

```js
/* eslint-disable quotes */
console.log("foo");
/* eslint-enable quotes */
```

That now throws an error on `console.log` as it violates the config quotes rule, even though it's disabled for that block.
